internal/dbtools/notification_preferences.go

```diff
@@ -0,0 +1,303 @@
+
+// UserNotificationPreferenceTargets is an alias for user notification target
+// slice
+type UserNotificationPreferenceTargets []*UserNotificationPreferenceTarget
+
+// UserNotificationPreference is the user notification preference response
+type UserNotificationPreference struct {
+	NotificationType string `json:"notification_type" boil:"notification_type"`
+	Enabled          bool   `json:"enabled"`
+
+	NotificationTargets UserNotificationPreferenceTargets `json:"notification_targets"`
+}
+
+// UserNotificationPreferences is an alias for user notification
+// preference slice
+type UserNotificationPreferences []*UserNotificationPreference
+
+// RefreshNotificationDefaults refreshes the notification_defaults
+// materialized view
+func RefreshNotificationDefaults(ctx context.Context, ex boil.ContextExecutor) (err error) {
+	q := queries.Raw("REFRESH MATERIALIZED VIEW notification_defaults")
+	_, err = q.ExecContext(ctx, ex)
+
+	return
+}
+
+// GetNotificationPreferences fetch a user's notification preferences from
+// the governor DB, this function only gets rules the user had modified if
+// `withDefaults` is false
+func GetNotificationPreferences(ctx context.Context, uid string, ex boil.ContextExecutor, withDefaults bool) (UserNotificationPreferences, error) {
+	type preferencesQueryRecord struct {
+		NotificationType        string          `boil:"notification_type"`
+		NotificationTargetsJSON json.RawMessage `boil:"notification_targets"`
+	}
+
+	type preferencesQueryRecordNotificationTarget struct {
+		Target  string `json:"f1"`
+		Enabled bool   `json:"f2"`
+	}
+
+	queryMods := []qm.QueryMod{
+		qm.With(
+			`np as (
+				SELECT 
+					user_id,
+					notification_types.id as type_id,
+					notification_types.slug as type_slug,
+					notification_preferences.notification_target_id_null_string as target_id,
+					notification_targets.slug as target_slug,
+					enabled
+				FROM notification_preferences
+				LEFT JOIN notification_targets ON notification_target_id = notification_targets.id
+				LEFT JOIN notification_types ON notification_type_id = notification_types.id
+				WHERE notification_preferences.user_id = ?
+			)`,
+			uid,
+		),
+	}
+
+	if withDefaults {
+		queryMods = append(queryMods, qm.Select("nd.type_slug AS notification_type"))
+		queryMods = append(queryMods, qm.Select("jsonb_agg((nd.target_slug, IFNULL(np.enabled, nd.default_enabled))) AS notification_targets"))
+		queryMods = append(queryMods, qm.From("notification_defaults as nd"))
+		queryMods = append(queryMods, qm.FullOuterJoin("np on (np.target_id = nd.target_id AND np.type_id = nd.type_id)"))
+		queryMods = append(queryMods, qm.GroupBy("nd.type_slug"))
+	} else {
+		queryMods = append(queryMods, qm.Select("np.type_slug AS notification_type"))
+		queryMods = append(queryMods, qm.Select("jsonb_agg((np.target_slug, np.enabled)) AS notification_targets"))
+		queryMods = append(queryMods, qm.From("np"))
+		queryMods = append(queryMods, qm.GroupBy("np.type_slug"))
+	}
+
```

Comment:
Would it be better to create static raw queries like was done in hierarchical groups for both "withDefaults" and not? This is pretty complicated :laughing:

Reply:
another thought - would it be more straight forward to do the query for the user and if there were no results, return the default?

Reply:
yes a raw query looks much better.

we can't do this with the current data model where a user might have only one `type-target` config and the rest is from the `defaults` view